File: scripts/generate_audio.py
#!/usr/bin/env python3
"""Zero-credentials Azerbaijani audio generation using gTTS (Google Translate).

This is the easiest way to produce the bundled audio: no API key needed.
Quality is lower than Azure/Google Cloud neural voices, but it's perfectly
usable for a beginner learning pronunciation.

Reads data/cards.seed.json and writes public/audio/<cardId>.mp3 for every card
with Azerbaijani text (and hasAudio != false).

Setup:
    pip install gTTS
    python3 scripts/generate_audio.py            # skips files that exist
    python3 scripts/generate_audio.py --force    # regenerate everything

Note: gTTS calls Google Translate, so it needs internet access when you run it.
"""
import json
import os
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from _slug import az_slug  # noqa: E402

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SEED = os.path.join(ROOT, "data", "cards.seed.json")
LESSONS = os.path.join(ROOT, "data", "lessons.json")
OUT_DIR = os.path.join(ROOT, "public", "audio")

FORCE = "--force" in sys.argv


def lesson_jobs(seen):
    jobs = []
    try:
        with open(LESSONS, encoding="utf-8") as f:
            lessons = json.load(f)
    except FileNotFoundError:
        return jobs
    for lsn in lessons:
        for sec in lsn.get("theory", []):
            for a in sec.get("audio", []):
                name = "lsn-" + az_slug(a["az"])
                if name not in seen:
                    seen.add(name)
                    jobs.append((name, a["az"]))
    return jobs


def main() -> int:
    try:
        from gtts import gTTS
    except ImportError:
        print("gTTS is not installed. Run:  pip install gTTS", file=sys.stderr)
        return 1

    with open(SEED, encoding="utf-8") as f:
        seed = json.load(f)

    os.makedirs(OUT_DIR, exist_ok=True)
    # One clip per Azerbaijani text, plus one per example sentence (<id>.ex.mp3).
    jobs = []
    for c in seed["cards"]:
        if c.get("hasAudio") is False:
            continue
        if c.get("az"):
            jobs.append((c["id"], c["az"]))
        if c.get("ex"):
            jobs.append((f"{c['id']}.ex", c["ex"]))
    jobs += lesson_jobs({name for name, _ in jobs})
    print(f"gTTS · {len(jobs)} clips · out: {OUT_DIR}")

    made = skipped = failed = 0
    for name, text in jobs:
        out = os.path.join(OUT_DIR, f"{name}.mp3")
        if not FORCE and os.path.exists(out) and os.path.getsize(out) > 0:
            skipped += 1
            continue
        # Try Azerbaijani; Google Translate occasionally rate-limits, so retry
        # a few times with backoff. Keep going on persistent failure so one bad
        # word never blocks the whole (CI) run.
        ok = False
        for attempt in range(3):
            try:
                gTTS(text=text, lang="az", lang_check=False).save(out)
                ok = True
                break
            except Exception as err:  # noqa: BLE001
                wait = 1.5 * (attempt + 1)
                print(f"  … retry {name} in {wait:.1f}s ({err})", file=sys.stderr)
                time.sleep(wait)
        if ok:
            made += 1
            print(f"  ✓ {name}  {text}")
            time.sleep(0.4)  # be gentle on the endpoint
        else:
            failed += 1
            # Drop any empty/partial file so the app treats it as "no audio".
            if os.path.exists(out) and os.path.getsize(out) == 0:
                os.remove(out)
            print(f"  ✗ {name}  {text} (giving up)", file=sys.stderr)

    print(f"Done. {made} generated, {skipped} already present, {failed} failed.")
    # Succeed as long as we produced most of the audio; a few misses are fine.
    return 1 if jobs and made == 0 and failed > 0 else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/generate-audio.mjs
#!/usr/bin/env node
/**
 * Generate the bundled Azerbaijani audio set from data/cards.seed.json.
 *
 * Writes public/audio/<cardId>.mp3 for every card that has Azerbaijani text
 * (and hasAudio !== false). The app plays these offline; nothing here runs on
 * the user's phone.
 *
 * Providers (set TTS_PROVIDER):
 *   azure  (default, best az-AZ quality) — needs AZURE_SPEECH_KEY, AZURE_SPEECH_REGION
 *   google                               — needs GOOGLE_TTS_KEY (API key)
 *
 * For a zero-credentials option, use the Python gTTS script instead:
 *   npm run audio:gtts
 *
 * Usage:
 *   AZURE_SPEECH_KEY=xxx AZURE_SPEECH_REGION=westeurope npm run audio
 *   npm run audio -- --force     # re-generate even if the mp3 already exists
 */
import { readFile, mkdir, writeFile, access } from 'node:fs/promises'
import { constants } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const ROOT = join(__dirname, '..')
const SEED = join(ROOT, 'data', 'cards.seed.json')
const OUT_DIR = join(ROOT, 'public', 'audio')

const FORCE = process.argv.includes('--force')
const PROVIDER = (process.env.TTS_PROVIDER || 'azure').toLowerCase()

// Azure neural voices for Azerbaijani (Republic).
const AZURE_VOICE = process.env.AZURE_VOICE || 'az-AZ-BabekNeural'
const GOOGLE_VOICE = process.env.GOOGLE_VOICE || 'az-AZ-Standard-A'

async function exists(p) {
  try {
    await access(p, constants.F_OK)
    return true
  } catch {
    return false
  }
}

function xmlEscape(s) {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
}

async function synthAzure(text) {
  const key = process.env.AZURE_SPEECH_KEY
  const region = process.env.AZURE_SPEECH_REGION
  if (!key || !region) {
    throw new Error('Set AZURE_SPEECH_KEY and AZURE_SPEECH_REGION (e.g. westeurope).')
  }
  const ssml = `<speak version='1.0' xml:lang='az-AZ'><voice xml:lang='az-AZ' name='${AZURE_VOICE}'>${xmlEscape(
    text,
  )}</voice></speak>`
  const res = await fetch(`https://${region}.tts.speech.microsoft.com/cognitiveservices/v1`, {
    method: 'POST',
    headers: {
      'Ocp-Apim-Subscription-Key': key,
      'Content-Type': 'application/ssml+xml',
      'X-Microsoft-OutputFormat': 'audio-24khz-48kbitrate-mono-mp3',
      'User-Agent': 'azeri-for-dad',
    },
    body: ssml,
  })
  if (!res.ok) throw new Error(`Azure TTS ${res.status}: ${await res.text()}`)
  return Buffer.from(await res.arrayBuffer())
}

async function synthGoogle(text) {
  const key = process.env.GOOGLE_TTS_KEY
  if (!key) throw new Error('Set GOOGLE_TTS_KEY (a Google Cloud TTS API key).')
  const res = await fetch(`https://texttospeech.googleapis.com/v1/text:synthesize?key=${key}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      input: { text },
      voice: { languageCode: 'az-AZ', name: GOOGLE_VOICE },
      audioConfig: { audioEncoding: 'MP3' },
    }),
  })
  if (!res.ok) throw new Error(`Google TTS ${res.status}: ${await res.text()}`)
  const data = await res.json()
  return Buffer.from(data.audioContent, 'base64')
}

const synth = PROVIDER === 'google' ? synthGoogle : synthAzure

async function main() {
  const seed = JSON.parse(await readFile(SEED, 'utf8'))
  await mkdir(OUT_DIR, { recursive: true })

  const targets = seed.cards.filter((c) => c.az && c.hasAudio !== false)
  console.log(`Provider: ${PROVIDER} · ${targets.length} cards · out: ${OUT_DIR}`)

  let made = 0
  let skipped = 0
  for (const card of targets) {
    const out = join(OUT_DIR, `${card.id}.mp3`)
    if (!FORCE && (await exists(out))) {
      skipped++
      continue
    }
    try {
      const buf = await synth(card.az)
      await writeFile(out, buf)
      made++
      console.log(`  ✓ ${card.id}  ${card.az}`)
    } catch (err) {
      console.error(`  ✗ ${card.id}  ${card.az}\n    ${err.message}`)
      process.exitCode = 1
      return
    }
  }
  console.log(`Done. ${made} generated, ${skipped} already present.`)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
